eeps/eep-0065.md

```diff
@@ -0,0 +1,282 @@
+* «`c`»: char list, an Erlang string (Unicode codepoint list)
+* «`r`»: regular expression, pre-compiled.
+  [Sigil Suffix][]es such as in Elixir: «`u`», «`i`», «`s`», «`m`»,
+  «`x`», «`f`» and «`U`», should be implemented as flags to
+  the regular expression compilation.  Since Erlang is in the process of
+  changing its regular expression engine, these flags need an overhaul.
+
+All those Sigil Types allow both escape sequences and string interpolation.
+
+The capital letter Sigil Types «`S`», «`C`» and «`R`» are siblings
+to the above but takes the String Content verbatim as in
+does not handle escape sequences nor string interpolation.
+
+#### The Empty Sigil
+
+The Empty Sigil (default type) should either be disallowed, like in Elixir,
+or implement the most desired use case, which might be one of
+
+* Shorthand for today's binary strings:
+  «`~"Content"`» equivalent to «`<<"Content"/utf8>>`»
+* Regular strings with choice of [String Delimiters][]:
+  «`~{"Content"}`» equivalent to «`"\"Content\""`»
+* Equivalent to «`~s`»
+* Equivalent to «`~c`»
+
+The first two can be implemented today without deciding on
+the matter of string interpolation.
+
+### Comparison with Elixir
+
+The [Empty Sigil][] is not allowed in Elixir.
+
+All suggested [Sigil Type][]s in this proposal: «`s`», «`S`», «`c`», «`C`»,
+«`r`», and «`R`», would behave the same in both Erlang and Elixir
+with respect to escape sequences and string interpolation.  There is,
+however, small differences in which escape sequences that are implemented
+in the languages; Elixir allows escaping of newlines, and has got
+an escape sequence «`\a`», which Erlang does not.
+
+There are also small differences in how newlines are handled
+between heredocs in Elixir and triple-quoted strings in Erlang.
+See [EEP 64][].
+
+In [Elixir][1], adding a «`~s`» sigil to a regular string or a heredoc
+creates an equivalent string.
+
+According to this suggestion, since strings are character lists
+in Erlang, instead adding a «`~c`» sigil to a regular string
+or a triple-quoted string, would create an equivalent string
+*except* that the sigil string should implement string interpolation.
+
+Details about regular expression sigils, «`~r`» and «`~R`», in particular
+their [Sigil Suffix][]es remains to be decided, so they might differ.
+
+It has not been decided how or even *if* string interpolation will
+be implemented in Erlang, so it surely might differ.
+
+Reference Implementation
+------------------------
+
+[PR-7684][] Implements the basics of handling Sigils on string literals.
+The scanner produces a «`sigil`» token before the string literal, and a
+«`sigil_suffix`» token after.  These are then removed by the parser.
+
+Another approach would be to produce e.g a «`sigil_string`» token
+for the whole string and then handle that in the parser.
+It would require more state to be kept in the scanner between
+the parts of the sigil prefixed string, and therefore need
+more scanner rewriting.
```

Comment:
The reason why Elixir introduces lower-cased and upper-cased sigils is because you can import custom functions as sigils. `~FOO["bar"]` is equivalent to calling `sigil_FOO("bar")`. Therefore, because interpolation is a tokenizer/parser feature, the name needs to tell us if the feature is enabled or disabled.

On the Erlang side, you could implement this by emitting new AST nodes. This means someone can use a parse transform (or something else) to give meaning to new sigils.

However, if the feature is fully implemented on the parser/tokenizer and is not extensible, then you know upfront if a sigil supports interpolation or escape codes. This means you don't need to rely on the lower/upper case sigil rule. You could use `unescaped_string"foo"` or even introduce `s"foo"U` or some other letter to mean escaping is disabled (although having such important information at the end of the sigil is weird).

Reply:
I think we want to implement this feature in the lexer+parser and not have it extendable, other than through parse transforms.

It would be natural for the lexer to remember the sigil prefix when scanning the string, so it then knows how to handle escape codes (and interpolation).  If that information would be in the suffix, the scanner would have to first do a raw scan of the string, and then re-scan it if the suffix says that escape codes should be honoured.  That would be awkward and disturb error reporting since e.g a string end error would be detected before an escape sequence error that is located before the string end in the code.

So seems better if it is only the sigil prefix that affects how the lexer should scan the string content.

You are right in that there is no reason to have Elixir's uppercase/lowercase convention for escape code and interpolation scanning in Erlang, other than for having similar look and feel between the languages, when possible.  

Another approach could be to interpret the prefix as characters and have the lexer recognize some character for escape codes, interpolation, and whatnot...  But if we bundle escape codes + interpolation in the same flag, then the uppercase/lowercase convention works.